[mod] searx.engines.__init__: refactoring

dalf opened the pull request:
## What does this PR do?

See comments

## Why is this change important?

<!-- MANDATORY -->

<!-- explain the motivation behind your PR -->

## How to test this PR locally?

<!-- commands to run the tests or instructions to test the changes-->

## Author's checklist

<!-- additional notes for reviewiers -->

## Related issues

<!--
Closes #234
-->

dalf committed:
[mod] searx.engines.__init__: refactoring

dalf commented on searx/engines/__init__.py:
```diff
@@ -67,61 +53,57 @@ def load_engine(engine_data):
+        setattr(engine, 'supported_languages', ENGINES_LANGUAGES[engine_name])
 
     # find custom aliases for non standard language codes
-    if hasattr(engine, 'supported_languages'):
```
The condition is removed because it is always True: `engine_default_args` contains `supported_languages`, so the ` for arg_name, arg_value in engine_default_args.items():...` set the attribute `supported_languages`.

dalf commented on searx/engines/__init__.py:
```diff
@@ -67,61 +53,57 @@ def load_engine(engine_data):
         if engine_attr.startswith('_'):
             continue
-        if engine_attr == 'inactive' and getattr(engine, engine_attr) is True:
-            return None
```
Moved above.

dalf commented on searx/engines/__init__.py:
```diff
@@ -67,61 +53,57 @@ def load_engine(engine_data):
     for arg_name, arg_value in engine_default_args.items():
         if not hasattr(engine, arg_name):
-            setattr(engine, arg_name, arg_value)
+            setattr(engine, arg_name, copy.deepcopy(arg_value))
```
deepcopy because `language_aliases` is going to be modified (should be an actual comment)

dalf commented on searx/engines/__init__.py:
```diff
@@ -146,24 +128,27 @@ def load_engine(engine_data):
 
 
 def load_engines(engine_list):
-    global engines, engine_shortcuts
```
The references are kept, so there is no write access, so there is no need for `global`.

return42 commented on searx/engines/__init__.py:
```diff
@@ -45,6 +28,7 @@
 engine_default_args = {'paging': False,
                        'categories': ['general'],
```
just a nit: while refactoring, lets fix indentation here  

    engine_default_args = {
        'paging': False,
        ....

return42 commented on searx/engines/__init__.py:
```diff
@@ -57,6 +41,8 @@
 
 
 def load_engine(engine_data):
+    # pylint: disable=too-many-branches, too-many-statements, protected-access
```
Do you like to add a doc-string that documents 

- engine attributes from YAML `settings:engines/<engine>` argument `engine_data`
- set default attributes (from `engine_default_args`)
- check _inactive_  is True or False
- what does _"checking required variables"_ means

return42 commented on searx/engines/__init__.py:
```diff
@@ -67,61 +53,57 @@ def load_engine(engine_data):
-            pass
-        elif param_name == 'categories':
+        if param_name == 'categories':
             if param_value == 'none':
```
Do we support category "none" .. I haven't found any usage of a string "none" in settings.yml nor in one of the `engines/*.py`

return42 committed:
[mod] searx.engines.load_engine return None instead of sys.exit(1)

Loading an engine should not exit the application (*). Instead
of exit, return None.

(*) RuntimeError still exit the application: syntax error, etc...

BTW: add documentation and normalize indentation (no functional change)

Suggested-by: @dalf https://github.com/searxng/searxng/pull/116#issuecomment-851865627
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[docs] add new rubric 'Source-Code'

- document namespace: searx.engines
- move docs/dev/xpath_engine.rst -> docs/src/searx.engines.xpath.rst

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[docs] highlight source code linked from *automodule* content

sphinx.ext.viewcode won't highlight when 'highlight_language' [1] is set to
string 'none' [2]

[1] https://www.sphinx-doc.org/en/master/usage/extensions/viewcode.html
[2] https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-highlight_language

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 reviewed (approved):
@dalf I pushed three commits on top of your branch / except 1c13beee no functional change

If you think this commits have a value just merge them along this PR into master, otherwise drop what is not needed.

return42 commented on searx/engines/__init__.py:
```diff
@@ -1,169 +1,220 @@
+    engines[engine.name] = engine
     if engine.shortcut in engine_shortcuts:
         logger.error('Engine config error: ambigious shortcut: {0}'.format(engine.shortcut))
         sys.exit(1)
```
Should we remove this sys.exit() also?

dalf replied:
I'm not sure what should be done, so we can keep sys.exit(1).

But the name has the same issue with the name:
```python
    if engine.name in engines:
        logger.error('Engine config error: ambigious name: {0}'.format(engine.name))
        sys.exit(1)

    if engine.shortcut in engine_shortcuts:
        logger.error('Engine config error: ambigious shortcut: {0}'.format(engine.shortcut))
        sys.exit(1)
```

dalf committed:
[fix] sys.exit(1) when there is duplicate engine name

return42 reviewed (approved):
Thanks! .. lets merge ..

dalf commented on searx/engines/__init__.py:
```diff
@@ -1,169 +1,225 @@
+        engine = load_module(engine_module + '.py', ENGINE_DIR)
     except (SyntaxError, KeyboardInterrupt, SystemExit, SystemError, ImportError, RuntimeError):
         logger.exception('Fatal exception in engine "{}"'.format(engine_module))
         sys.exit(1)
```
Still one `sys.exit(1)` here otherwise it is possible to have a syntax error in the engine code (it had happened before).
